lib/signs/utilities/audio.ex

```diff
@@ -11,207 +11,220 @@ defmodule Signs.Utilities.Audio do
+
+              {Audio.Approaching.from_message(message, include_crowding?),
+               sign
+               |> update_in(
+                 [Access.key!(:announced_approachings)],
+                 &cache_value(&1, message.trip_id)
+               )
+               |> update_in(
+                 [Access.key!(:announced_approachings_with_crowding)],
+                 &if(include_crowding?, do: cache_value(&1, message.trip_id), else: &1)
+               )}
+
+            match?(%Message.StoppedTrain{}, message) && index == 0 &&
+                {message.trip_id, message.stops_away} not in sign.announced_stalls ->
+              {Audio.StoppedTrain.from_message(message),
+               update_in(
+                 sign.announced_stalls,
+                 &cache_value(&1, {message.trip_id, message.stops_away})
+               )}
+
+            match?(%Message.Predictions{}, message) && is_integer(message.minutes) && index == 0 &&
+              sign.prev_prediction_keys &&
+                {message.route_id, message.direction_id} not in sign.prev_prediction_keys ->
+              {Audio.NextTrainCountdown.from_message(message), sign}
+
+            true ->
+              {[], sign}
           end
-        end
-      )
+        end)
+      end)
 
-    new_announced_approaching_with_crowding =
-      for %Audio.Approaching{trip_id: trip_id, crowding_description: crowding_description} <-
-            new_audios,
-          not is_nil(trip_id) and not is_nil(crowding_description) do
-        trip_id
-      end
+    log_crowding(new_audios, sign.id)
 
     sign = %{
       sign
-      | announced_approachings: Enum.take(new_approaching_trips, @announced_history_length),
-        announced_approachings_with_crowding:
-          Enum.take(
-            new_announced_approaching_with_crowding ++ sign.announced_approachings_with_crowding,
-            @announced_history_length
-          ),
-        announced_arrivals: Enum.take(new_arriving_trips, @announced_history_length)
+      | prev_prediction_keys:
+          for {:predictions, list} <- items, message <- list, uniq: true do
+            {message.route_id, message.direction_id}
+          end
     }
 
+    # Disable crowding messages for now
     new_audios =
-      if SourceConfig.multi_source?(sign.source_config) do
-        sort_audio(new_audios)
-      else
-        new_audios
-      end
-      |> tap(&log_crowding(&1, sign.id))
-      |> Enum.map(fn %{__struct__: audio_type} = audio ->
+      Enum.map(new_audios, fn %{__struct__: audio_type} = audio ->
         if audio_type in [Audio.Approaching, Audio.TrainIsArriving] and sign.id not in [] do
           %{audio | crowding_description: nil}
         else
           audio
         end
       end)
 
-    {new_audios, sign}
+    {audios ++ new_audios, sign}
+  end
+
+  defp cache_value(list, value), do: [value | list] |> Enum.take(@announced_history_length)
+
+  defp decode_sign(sign) do
```

Comment:
Note: This function provides a sneak peek into where this refactor is headed. It's essentially "decompiling" the granular line-oriented messages into higher-level semantic structures, which are more suitable for transforming and generating e.g. audio messages. Eventually, the goal is to make proper types for these, and create them earlier in the process, so the semantic forms can be used more naturally, without having to reverse engineer them.

Reply:
Clarifying question - when you say "make proper types", do you just mean to more explicitly define them using typespecs somewhere or do you mean something else?

Reply:
I meant creating named structs. The data here is already kind of leaning in that direction (with the e.g. `{:predictions, ...}` shapes), but since the scope of this data is very limited to start, I didn't want to jump ahead and make it concrete before we could see more of the use cases.

Reply:
Right - considering structs are just maps with a `__struct__` key, you could say these semantic structures are almost there. So the idea is to codify our content types at the high level with named structs, but eliminate the use of types at the end of the pipeline to streamline the serialization process. Am I understanding that correctly?